org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java

```diff
@@ -596,6 +618,79 @@ protected Compiler newCompiler() {
 	return newCompiler;
 }
 
+private CompilerConfiguration prepareCompilerConfiguration(CompilerOptions options) {
+	try {
+		List<URI> annotationProcessorPaths = new ArrayList<>();
+		List<IContainer> generatedSourcePaths = new ArrayList<>();
+		boolean isTest = this.compilationGroup == CompilationGroup.TEST;
+		if (this.javaBuilder.participants != null) {
+			for (CompilationParticipant participant : this.javaBuilder.participants) {
+				if (participant.isAnnotationProcessor()) {
+					URI[] paths = participant.getAnnotationProcessorPaths(this.javaBuilder.javaProject, isTest);
+					if (paths != null) {
+						annotationProcessorPaths.addAll(Arrays.asList(paths));
+					}
+					IContainer[] generatedSrc = participant.getGeneratedSourcePaths(this.javaBuilder.javaProject, isTest);
+					if (generatedSrc != null) {
+						generatedSourcePaths.addAll(Arrays.asList(generatedSrc));
+					}
+				}
+			}
+		}
+
+		ClasspathLocation[] classpathLocations = this.nameEnvironment.binaryLocations;
+		Set<URI> classpaths = new LinkedHashSet<>();
+		Set<URI> modulepaths = new LinkedHashSet<>();
+		for (ClasspathLocation location : classpathLocations) {
+			if (location instanceof ClasspathDirectory cpDirectory) {
+				URI cpURI = cpDirectory.binaryFolder.getRawLocationURI();
+				if (cpURI == null) {
+					continue;
+				}
+				if (cpDirectory.isOnModulePath) {
+					modulepaths.add(cpURI);
+				} else {
+					classpaths.add(cpURI);
+				}
+			} else if (location instanceof ClasspathJar cpJar) {
+				URI cpURI = new File(cpJar.zipFilename).toURI();
```

Comment:
just one more thing: please avoid File.toURI but use implementation like org.eclipse.jdt.internal.core.search.indexing.BinaryIndexer.toUri(String) to avoid FileSystem access  - it is known that this file is a file and not a directory

Reply:
Ok, I will reuse the utility `org.eclipse.core.filesystem.URIUtil.toURI(filePath, true)`, which requires no IO operation.

Reply:
Cool, i did not know that method.